File: Cpp/K12.cpp
#include "K12.h"

#include <iostream>
#include <cmath>
#include <random>
#include <string>

#include <cstdio>
#include <ctime>

struct Person
{
	std::string name;
	int birthYear;
	float height;
	float weight;
};

struct Staff
{
	std::string name;
	char gender;
	int age;
};

struct Position
{
	int x;
	int y;
	double distance;
};

void K12::question1()
{
	Person person;
	std::cout << "名前を入力してください＞";
	std::cin.clear();
	std::cin.ignore();
	std::getline(std::cin, person.name, '\n');
	std::cout << "生年（西暦）を入力してください＞";
	std::cin >> person.birthYear;
	std::cout << "身長を入力してください＞";
	std::cin >> person.height;
	std::cout << "体重を入力してください＞";
	std::cin >> person.weight;

	std::cout << "プロファイル：\n";
	std::cout << "名前：" << person.name << "\n";
	std::cout << "生年：" << person.birthYear << "\n";
	std::cout << "身長：" << person.height << "\n";
	std::cout << "体重：" << person.weight << "\n";
}

void K12::question2()
{
	Staff staffs[3] = {
		{"神木隆之介", 'M', 23},
		{"上白石萌音", 'F', 18},
		{"長澤まさみ", 'F', 28}
	};

	std::cout << "名前\t\t性別\t年齢\n";
	for (int i = 0; i <= 35; ++i)
	{
		std::cout << '-';
	}
	std::cout << "\n";

	for (const auto& staff : staffs)
	{
		std::cout << staff.name << "\t" << staff.gender << "\t" << staff.age << "歳\n";
	}
}

double getDistance(const Position& playerPos, const Position& enemyPos)
{
	int diffX = playerPos.x - enemyPos.x;
	int diffY = playerPos.y - enemyPos.y;
	return sqrt(diffX * diffX + diffY * diffY);
}

void K12::question3()
{
	std::mt19937 engine(static_cast<std::mt19937::result_type>(time(nullptr)));
	std::uniform_int_distribution<> rand(0, 101);
	auto getRandomNumber = [&]() { return rand(engine) - 50; };

	Position playerPos = {
		getRandomNumber(), getRandomNumber(), 0.0
	};
	Position enemyPos[5] = {};

	for (auto& enemyPo : enemyPos)
	{
		enemyPo.x = getRandomNumber();
		enemyPo.y = getRandomNumber();
		enemyPo.distance = getDistance(playerPos, enemyPo);
	}

	std::cout.setf(std::ios_base::fixed, std::ios_base::floatfield);
	std::cout.precision(2);
	printf("プレーヤーの座標：x= %.2f\t\ty= %.2f\n\n",
		static_cast<float>(playerPos.x), static_cast<float>(playerPos.y));
	for (int i = 0; i < 5; ++i)
	{
		printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離=%5.2f\n",
			i, static_cast<float>(enemyPos[i].x),
			static_cast<float>(enemyPos[i].y), static_cast<float>(enemyPos[i].distance));
	}
	
}

void swapPos(Position& a, Position& b)
{
	Position temp = a;
	a = b;
	b = temp;
}

void sortPos(Position* positions, int total)
{
	for (int i = 0; i < total; ++i)
	{
		for (int j = 0; j < total; ++j)
		{
			if (positions[j].distance > positions[i].distance)
			{
				swapPos(positions[i], positions[j]);
			}
		}
	}
}

void K12::question4()
{
	std::mt19937 engine(static_cast<std::mt19937::result_type>(time(nullptr)));
	std::uniform_int_distribution<> rand(0, 101);
	auto getRandomNumber = [&]() { return rand(engine) - 50; };

	Position playerPos = {
		getRandomNumber(), getRandomNumber(), 0.0
	};
	Position enemyPos[5] = {};

	for (auto& enemyPo : enemyPos)
	{
		enemyPo.x = getRandomNumber();
		enemyPo.y = getRandomNumber();
		enemyPo.distance = getDistance(playerPos, enemyPo);
	}

	std::cout.setf(std::ios_base::fixed, std::ios_base::floatfield);
	std::cout.precision(2);
	printf("プレーヤーの座標：x= %.2f\t\ty= %.2f\n\n",
		static_cast<float>(playerPos.x), static_cast<float>(playerPos.y));
	for (int i = 0; i < 5; ++i)
	{
		printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離=%5.2f\n",
			i, static_cast<float>(enemyPos[i].x),
			static_cast<float>(enemyPos[i].y), static_cast<float>(enemyPos[i].distance));
	}

	std::cout << "\n並び替え後\n";
	sortPos(enemyPos, 5);
	for (int i = 0; i < 5; ++i)
	{
		printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離=%5.2f\n",
			i, static_cast<float>(enemyPos[i].x),
			static_cast<float>(enemyPos[i].y), static_cast<float>(enemyPos[i].distance));
	}

}

File: Go/K12.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
	"strconv"
)

type K12 struct {}

type Person struct {
	Name string
	BirthYear int
	Height float32
	Weight float32
}

type Staff struct {
	Name string
	Gender uint8
	Age int
}

func newStaff(name string, gender uint8, age int) *Staff {
	_staff := new(Staff)
	_staff.Name = name
	_staff.Gender = gender
	_staff.Age = age
	return _staff
}

type Position struct {
	X int
	Y int
	Distance float64
}

func (k12 K12) execute(num int) {
	switch num {
	case 1: k12.question1()
	case 2: k12.question2()
	case 3: k12.question3()
	case 4: k12.question4()
	}
}

func (k12 K12) question1() {
	var person Person
	fmt.Print("名前を入力してください＞")
	_, _ = fmt.Scanln(&person.Name)
	fmt.Print("生年（西暦）を入力してください＞")
	var reader string
	_, _ = fmt.Scanln(&reader)
	person.BirthYear, _ = strconv.Atoi(reader)
	fmt.Print("身長を入力してください＞")
	_, _ = fmt.Scanln(&reader)
	num, _ := strconv.ParseFloat(reader, 32)
	person.Height = float32(num)
	fmt.Print("体重を入力してください＞")
	_, _ = fmt.Scanln(&reader)
	num, _ = strconv.ParseFloat(reader, 32)
	person.Weight = float32(num)

	fmt.Println("プロファイル")
	fmt.Printf("名前：%s\n", person.Name)
	fmt.Printf("生年：%d\n", person.BirthYear)
	fmt.Printf("身長：%.1f\n", person.Height)
	fmt.Printf("体重：%.1f\n", person.Weight)
}

func (k12 K12) question2() {
	staffs := [3]*Staff{
		newStaff("神木隆之介", 'M', 23),
		newStaff("上白石萌音", 'F', 18),
		newStaff("長澤まさみ", 'F', 28),
	}

	fmt.Println("名前\t\t性別\t年齢")
	for i := 0; i < 35; i++ {
		fmt.Print("-")
	}
	fmt.Println()

	for i := 0; i < len(staffs); i++ {
		fmt.Printf("%s\t%c\t%d歳\n", staffs[i].Name, staffs[i].Gender, staffs[i].Age)
	}
}

func getDistance(playerPos Position, enemyPos Position) float64 {
	diffX := playerPos.X - enemyPos.X
	diffY := playerPos.Y - enemyPos.Y
	return math.Sqrt(float64(diffX * diffX + diffY * diffY))
}

func (k12 K12) question3() {
	getRandomNumber := func() int {
		return rand.Intn(101) - 50
	}

	playerPos := Position{
		getRandomNumber(),
		getRandomNumber(),
		0.0,
	}

	var enemyPos []Position
	for i := 0; i < 5; i++ {
		pos := Position{
			getRandomNumber(),
			getRandomNumber(),
			0.0,
		}
		pos.Distance = getDistance(playerPos, pos)
		enemyPos = append(enemyPos, pos)
	}

	fmt.Printf("プレーヤーの座標：x= %d\t\ty= %d\n", playerPos.X, playerPos.Y)
	for i := 0; i < len(enemyPos); i++ {
		fmt.Printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f\n",
			i, float32(enemyPos[i].X), float32(enemyPos[i].Y), enemyPos[i].Distance)
	}
}

func (k12 K12) question4() {
	getRandomNumber := func() int {
		return rand.Intn(101) - 50
	}

	playerPos := Position{
		getRandomNumber(),
		getRandomNumber(),
		0.0,
	}

	var enemyPos []Position
	for i := 0; i < 5; i++ {
		pos := Position{
			getRandomNumber(),
			getRandomNumber(),
			0.0,
		}
		pos.Distance = getDistance(playerPos, pos)
		enemyPos = append(enemyPos, pos)
	}

	fmt.Printf("プレーヤーの座標：x= %d\t\ty= %d\n", playerPos.X, playerPos.Y)
	for i := 0; i < len(enemyPos); i++ {
		fmt.Printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f\n",
			i, float32(enemyPos[i].X), float32(enemyPos[i].Y), enemyPos[i].Distance)
	}

	sort.Slice(enemyPos, func(i, j int) bool {
		return enemyPos[i].Distance < enemyPos[j].Distance
	})

	fmt.Println("\n並び替え後")
	for i := 0; i < len(enemyPos); i++ {
		fmt.Printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f\n",
			i, float32(enemyPos[i].X), float32(enemyPos[i].Y), enemyPos[i].Distance)
	}
}

